[WebGPUSwift] Replace copyBufferToTexture with a Swift implementation

nmahendru opened the pull request:
#### 9aa7a3270fbb68cbecf8e05ea841447848cc3885
<pre>
[WebGPUSwift] Replace copyBufferToTexture with a Swift implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=281757">https://bugs.webkit.org/show_bug.cgi?id=281757</a>
<a href="https://rdar.apple.com/138188859">rdar://138188859</a>

Reviewed by Mike Wyrzykowski.

The replacement is 1-for-1. Disabled by default.
One thing which is not the best is that we had to do a lot of Integer
conversions which are always explicit in Swift but implicit in C++.
NSUInteger has been mapped to UInt.

The return values of APIConversions.h::fromAPI are usable in Swift and Swift should
do its counting to hold a strong reference until it needs that value. So there is no
need for Ref&lt;Buffer&gt;(apiBuffer) and Ref&lt;Texture&gt;(destinationTexture) as used in the C++ Implementation.

Fly by changes in this PR:
1. Swift 6 mode is now enabled.
2. The duplicate definitions of retain/release helpers for the C++ refcounted classes have been cleaned up.

Some workarounds:
1. &lt;pthread/private.h&gt; has been imported without extern &quot;C&quot; to facilitate usage with Swift Clang Importer.
2. &lt;os/lock_private.h&gt; needs OS_UNFAIR_LOCK_INLINE=1 but as it&apos;s being imported as a module for the swift build
   the define needs to be passed in as a command line flag as modules are precompiled and they don&apos;t consider
   what&apos;s including them and what&apos;s defined above them.
   (The assumption is that WebGPUSwift build will always use the internal SDK)

Testing:
Manually ran the test:
./LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/image_copy.html

* Source/WTF/wtf/RefCounted.h:
(WTF::ref):
(WTF::deref):
(WTF::retainRefCounted): Deleted.
(WTF::releaseRefCounted): Deleted.
* Source/WTF/wtf/RetainReleaseSwift.h:
* Source/WTF/wtf/ThreadSafeRefCounted.h:
(WTF::retainThreadSafeRefCounted): Deleted.
(WTF::releaseThreadSafeRefCounted): Deleted.
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::retainThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr): Deleted.
(WTF::releaseThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr): Deleted.
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU/Buffer.h:
(refBuffer):
(derefBuffer):
(retainBuffer): Deleted.
(releaseBuffer): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.h:
(refCommandEncoder):
(derefCommandEncoder):
(retainCommandEncoder): Deleted.
(releaseCommandEncoder): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_copyBufferToTexture_thunk(_:source:destination:copySize:)):
(WebGPU.copyBufferToTexture(_:destination:copySize:)):
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::limitsCopy const):
(refDevice):
(derefDevice):
(retainDevice): Deleted.
(releaseDevice): Deleted.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
* Source/WebGPU/WebGPU/QuerySet.h:
(refQuerySet):
(derefQuerySet):
(retainQuerySet): Deleted.
(releaseQuerySet): Deleted.
* Source/WebGPU/WebGPU/Queue.h:
(refQueue):
(derefQueue):
(retainQueue): Deleted.
(releaseQueue): Deleted.
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):
* Source/WebGPU/WebGPU/Texture.h:
(refTexture):
(derefTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h:
* Source/bmalloc/libpas/src/libpas/pas_lock.h:
* Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.h:
(pas_msl_malloc_logging):
(pas_msl_free_logging):

Canonical link: <a href="https://commits.webkit.org/287320@main">https://commits.webkit.org/287320@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/931b463cdc13cac1933b65432e4d85fd04a4c192

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79150 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58182 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32525 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83781 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30348 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67281 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6448 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61932 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/19845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82217 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72133 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42237 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49334 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26258 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28721 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72268 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26681 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85171 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78362 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6460 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70175 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68004 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69423 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13474 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12280 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100689 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12222 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6414 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/22618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/6361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9821 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/8153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.h:
```diff
@@ -117,6 +117,10 @@ class CommandEncoder : public WGPUCommandEncoderImpl, public RefCounted<CommandE
     bool validateClearBuffer(const Buffer&, uint64_t offset, uint64_t size);
     static void trackEncoder(CommandEncoder&, WeakHashSet<CommandEncoder>&);
+#if ENABLE(WEBGPU_SWIFT) && defined(__WEBGPU__)
+    uint32_t maxTextureDimension1D();
```
These could be const member functions I think

nmahendru replied:
I'll see if I can just make the m_device accesible to swift.

mwyrzykowski commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -177,12 +177,15 @@ WGPU_EXPORT void wgpuDeviceClearUncapturedErrorCallback(WGPUDevice device) WGPU_
 #define HAS_SWIFTCXX_THUNK  NS_REFINED_FOR_SWIFT
 
 #if defined(ENABLE_WEBGPU_SWIFT) && ENABLE_WEBGPU_SWIFT && defined(__WEBGPU__)
+#include <Metal/Metal.h>
```
why do we need to include these headers into the bottom of a header file?

nmahendru replied:
IIRC the headers need some stuff declared in WebGPUExt.h.

mwyrzykowski replied:
Shouldn't WebGPUExt.h be included in the headers then and not the other way around?

emw-apple replied:
Maybe we should move these to a `WebGPUSwift.h` header, to better separate intent.

I added one in https://commits.webkit.org/286661@main but then removed it in https://commits.webkit.org/286822@main, because I thought WebGPUExt.h was a better home for declarations that should be included widely inside the WebGPU target.

Still, including Metal in here doesn't make sense to me, since it's not Swift specific. Why is the import unavoidable, and if we truly need it, can the Metal import be moved to the top?

nmahendru replied:
Yea this header list is not in the right place and I am planning to clean this up in:
https://github.com/WebKit/WebKit/pull/36998
(Note that above PR is not complete yet so I will have another look once I am done with this one)

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -23,6 +23,11 @@
+import Metal
+
+private let WGPU_COPY_STRIDE_UNDEFINED = 0xffffffff
+private let MTLBlitOptionNone: UInt = 0
```
is there a swift equivalent? In Swift, the docs say MTLBlitOption is a struct - https://developer.apple.com/documentation/metal/mtlblitoption

nmahendru replied:
There isn't.It might be that MTLBiltOptionsNone is the default value but I just wanted to be sure.

geoffreygaren replied:
Can we import MTLBlitOptionNone from Metal instead of hard-coding it ourselves?

nmahendru replied:
It's not exposed in Swift. @geoffreygaren.
https://developer.apple.com/documentation/metal/mtlblitoption/mtlblitoptionnone

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -32,7 +37,314 @@ public func resolveQuerySet(commandEncoder: WebGPU.CommandEncoder, querySet: Web
+                    options: options
+                )
+            case WGPUTextureDimension_Force32.rawValue:
+                // FIXME: need debug assert here.
```
I don't think we need an assertion as this is validated earlier, `return` is fine

nmahendru replied:
Sure. Should I remove it from the ObjectiveC version as well then ? 
I honestly don't know what this code is doing so I am trying my best to convert it apples for apples.

geoffreygaren replied:
I'd suggest matching the ObjC code. assertionFailure() should do the trick.

nmahendru replied:
adding assertionFailure.

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -32,7 +37,314 @@ public func resolveQuerySet(commandEncoder: WebGPU.CommandEncoder, querySet: Web
+                //precondition(false, "ASSERT_NOT_REACHED")
+                return
+            default:
+                // FIXME: need debug assert here.
```
same as above

nmahendru replied:
ditto.

geoffreygaren replied:
Ditto

mwyrzykowski commented on Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.h:
```diff
@@ -64,7 +64,7 @@ static PAS_ALWAYS_INLINE bool pas_msl_is_enabled(void)
 static PAS_ALWAYS_INLINE pas_allocation_result pas_msl_malloc_logging(size_t size, pas_allocation_result result)
 {
-#if PAS_OS(DARWIN)
+#if PAS_OS(DARWIN) && !defined(__swift__) // FIXME: no malloc_logger for swift. see pas_darwin_spi.h
```
should the FIXME have a bug link?

emw-apple commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -32,7 +37,314 @@ public func resolveQuerySet(commandEncoder: WebGPU.CommandEncoder, querySet: Web
+        }
+        guard !(apiBuffer.isDestroyed() || destinationTexture.isDestroyed()) else {
+            return
+        }
```
I think the logic here is wrong. It should be something like this:

```suggestion
        guard copySize.width != 0 || copySize.height != 0 || copySize.depthOrArrayLayers != 0,
            !apiBuffer.isDestroyed(),
            !destinationTexture.isDestroyed()
            else {
                return
        }
```

I think combining into one guard clause makes it more understandable, too.

nmahendru replied:
good catch!. I will fix that.

emw-apple commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -32,7 +37,314 @@ public func resolveQuerySet(commandEncoder: WebGPU.CommandEncoder, querySet: Web
+        let aspectSpecificFormat = WebGPU.Texture.aspectSpecificFormat(destinationTexture.format(), destination.aspect)
+        let blockSize = WebGPU.Texture.texelBlockSize(aspectSpecificFormat)
+
+        switch destinationTexture.dimension().rawValue {
```
Why check rawValue instead of the typed value?

nmahendru replied:
That's the only way to write a switch of C++ types in swift as C++ enums are imported as Structs in swift.
https://www.swift.org/documentation/cxx-interop/#using-c-enumerations

emw-apple replied:
Switch statements aren't restricted to enums. They will match any two expressions that have a defined pattern-matching operator (`~=`). By default, any types that are Equatable can pattern match.

https://docs.swift.org/swift-book/documentation/the-swift-programming-language/patterns/#Expression-Pattern

I removed the `.rawValue`s locally and it compiles fine.

nmahendru replied:
ooo.. okie yea.. I think I translated `==` operator here. removing `rawValue` from switch statements.

emw-apple commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -32,7 +37,314 @@ public func resolveQuerySet(commandEncoder: WebGPU.CommandEncoder, querySet: Web
+        }
+
+        var options: MTLBlitOption = MTLBlitOption(rawValue: MTLBlitOptionNone)
+        switch destination.aspect.rawValue {
```
ditto

nmahendru replied:
ditto

emw-apple reviewed (commented):
The APIConversions.h functions (`fromAPI` and protectedFromAPI) feel very weird in a Swift context. I'm curious if you've thought a more idiomatic Swift alternative? Do we even need them?

emw-apple commented on Source/WTF/wtf/RefCounted.h:
```diff
@@ -214,17 +214,13 @@ template<typename T> class RefCounted : public RefCountedBase {
-    RELEASE_ASSERT(obj != nullptr);
-    static_assert(std::derived_from<T, WTF::RefCounted<T>>);
-    static_cast<WTF::RefCounted<T>*>(obj)->ref();
+    obj->ref();
```
Why this change? Modifications to WTF inlines like this without a clear explanation scare me, although it looks like these two functions are only used by WebGPU code.

nmahendru replied:
This was just a cleanup leftover from previous PRs. Geoff had some previous feedback which I am just working on here. 
https://github.com/WebKit/WebKit/pull/34305#discussion_r1779815225

emw-apple commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -35,7 +35,9 @@ public func writeBuffer(
 extension WebGPU.Queue {
     public func writeBuffer(_ buffer: WebGPU.Buffer, bufferOffset: UInt64, data: SpanUInt8) {
-        let device = self.device()
+        guard let device = self.m_device.get().get() else {
+            return
+        }
```
Why did this need to change?

nmahendru replied:
This change is based on Geoff's feedback on some other PR for a similar change I made.. I can't find that link for now..
In this PR itself, I have removed the `SWIFT_RETURNS_INDEPENDENT_VALUE` from `Queue::device()` as that was wrong in the first place. Swift by default does not allow vanilla C++ references and I had basically "made" it to by adding this annotation earlier. But it's making this api a sharp edge that can get a device reference that might not live as long as the Queue.

emw-apple commented on Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h:
```diff
@@ -31,9 +31,16 @@
+#if !(defined(__has_feature) && __has_feature(modules))
 PAS_END_EXTERN_C;
+#endif
+
```
This still feels really weird to me. I think we should conditionally disable the include entirely, rather than import it as extern-C in some places and as C++ in another.

Also, rdar://140431798 is a better radar to associate this with.

geoffreygaren replied:
Yeah, I tend to agree that we should just not #include at all if we can't #include correctly.

nmahendru replied:
I have fixed this by just including the modularized header in the newer sdk.

emw-apple commented on Source/bmalloc/libpas/src/libpas/pas_lock.h:
```diff
@@ -124,6 +124,14 @@ PAS_END_EXTERN_C;
+#define os_unfair_lock_trylock_inline os_unfair_lock_trylock
+#define os_unfair_lock_unlock_inline os_unfair_lock_unlock
+#endif
+
```
I worry if this is an ODR violation which could cause performance regressions. These functions may be used to generate multiple copies of the same instantiation in different files, and at link time, a copy generated for Swift compilation may be chosen. Then all paths through that template instantiation will use the slow path.

nmahendru replied:
fixed this by passing in OS_UNFAIR_LOCK_INLINE=1

geoffreygaren commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -23,6 +23,11 @@
 
+import Metal
+
+private let WGPU_COPY_STRIDE_UNDEFINED = 0xffffffff
```
Can we import this value from WebGPU.h?

If there's something stopping us from using a #define directly, can we change the #define to a static const, or add a using declaration for Swift like we do for templates?

geoffreygaren commented on Source/WebGPU/WebGPU/Texture.h:
```diff
@@ -162,6 +163,16 @@ class Texture : public WGPUTextureImpl, public RefCounted<Texture>, public CanMa
 
 } // namespace WebGPU
+
+inline void retainTexture(WebGPU::Texture* obj)
```
Let's call this "refTexture". "retain" is what we do with ObjC objects.

geoffreygaren commented on Source/WebGPU/WebGPU/Texture.h:
```diff
@@ -162,6 +163,16 @@ class Texture : public WGPUTextureImpl, public RefCounted<Texture>, public CanMa
+
+inline void retainTexture(WebGPU::Texture* obj)
+{
+    WTF::retainRefCounted(obj);
```
...and we can call this 'ref'. A free function that does `obj->ref()` is `ref(obj)`. Changing the argument positions doesn't need to change the name.

geoffreygaren commented on Source/WebGPU/WebGPU/Texture.h:
```diff
@@ -162,6 +163,16 @@ class Texture : public WGPUTextureImpl, public RefCounted<Texture>, public CanMa
+    WTF::retainRefCounted(obj);
+}
+
+inline void releaseTexture(WebGPU::Texture* obj)
```
derefTexture

geoffreygaren commented on Source/WebGPU/WebGPU/Texture.h:
```diff
@@ -162,6 +163,16 @@ class Texture : public WGPUTextureImpl, public RefCounted<Texture>, public CanMa
+
+inline void releaseTexture(WebGPU::Texture* obj)
+{
+    WTF::releaseRefCounted(obj);
```
deref(obj)

nmahendru replied:
Have implemented all your suggestions in this area.

nmahendru commented on Source/WTF/wtf/ThreadSafeRefCounted.h:
```diff
@@ -154,22 +154,6 @@ template<class T, DestructionThread destructionThread = DestructionThread::Any>
     ThreadSafeRefCounted() = default;
 };
 
-template<typename T>
```
Now using just one free function defined in RefCounted.h

nmahendru committed:
[WebGPUSwift] Replace copyBufferToTexture with a Swift implementation
https://bugs.webkit.org/show_bug.cgi?id=281757
rdar://138188859

Reviewed by Mike Wyrzykowski.

The replacement is 1-for-1. Disabled by default.
One thing which is not the best is that we had to do a lot of Integer
conversions which are always explicit in Swift but implicit in C++.
NSUInteger has been mapped to UInt.

The return values of APIConversions.h::fromAPI are usable in Swift and Swift should
do its counting to hold a strong reference until it needs that value. So there is no
need for Ref<Buffer>(apiBuffer) and Ref<Texture>(destinationTexture) as used in the C++ Implementation.

Fly by changes in this PR:
1. Swift 6 mode is now enabled.
2. The duplicate definitions of retain/release helpers for the C++ refcounted classes have been cleaned up.

Some workarounds:
1. <pthread/private.h> has been imported without extern "C" to facilitate usage with Swift Clang Importer.
2. <os/lock_private.h> needs OS_UNFAIR_LOCK_INLINE=1 but as it's being imported as a module for the swift build
   the define needs to be passed in as a command line flag as modules are precompiled and they don't consider
   what's including them and what's defined above them.
   (The assumption is that WebGPUSwift build will always use the internal SDK)

Testing:
Manually ran the test:
./LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/image_copy.html

* Source/WTF/wtf/RefCounted.h:
(WTF::ref):
(WTF::deref):
(WTF::retainRefCounted): Deleted.
(WTF::releaseRefCounted): Deleted.
* Source/WTF/wtf/RetainReleaseSwift.h:
* Source/WTF/wtf/ThreadSafeRefCounted.h:
(WTF::retainThreadSafeRefCounted): Deleted.
(WTF::releaseThreadSafeRefCounted): Deleted.
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::retainThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr): Deleted.
(WTF::releaseThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr): Deleted.
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU/Buffer.h:
(refBuffer):
(derefBuffer):
(retainBuffer): Deleted.
(releaseBuffer): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.h:
(refCommandEncoder):
(derefCommandEncoder):
(retainCommandEncoder): Deleted.
(releaseCommandEncoder): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_copyBufferToTexture_thunk(_:source:destination:copySize:)):
(WebGPU.copyBufferToTexture(_:destination:copySize:)):
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::limitsCopy const):
(refDevice):
(derefDevice):
(retainDevice): Deleted.
(releaseDevice): Deleted.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
* Source/WebGPU/WebGPU/QuerySet.h:
(refQuerySet):
(derefQuerySet):
(retainQuerySet): Deleted.
(releaseQuerySet): Deleted.
* Source/WebGPU/WebGPU/Queue.h:
(refQueue):
(derefQueue):
(retainQueue): Deleted.
(releaseQueue): Deleted.
* Source/WebGPU/WebGPU/Queue.swift:
(WebGPU.writeBuffer(_:bufferOffset:data:)):
* Source/WebGPU/WebGPU/Texture.h:
(refTexture):
(derefTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h:
* Source/bmalloc/libpas/src/libpas/pas_lock.h:
* Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.h:
(pas_msl_malloc_logging):
(pas_msl_free_logging):

Canonical link: https://commits.webkit.org/287320@main
